src/ATen/native/xpu/sycl/BucketizationKernels.cpp

```diff
@@ -0,0 +1,201 @@
+#include <ATen/ATen.h>
+#include <comm/SYCLContext.h>
+
+namespace at::native::xpu {
+
+// customized lower_bound func to ensure the low bound of 'nan', 'inf' etc. be
+// the end of boundary and we can properly handle a sorter argument
+// std::lower_bound can not be used here since its customized comparator need
+// strict weak ordering and the customized comparators require both arguments to
+// have the same type, which wouldn't happen when comparing val of input_t to an
+// indexer value from sorter of int64
+template <typename input_t>
+int64_t cus_lower_bound(
+    int64_t start,
+    int64_t end,
+    const input_t val,
+    const input_t* bd,
+    const int64_t* sort) {
+  // sorter gives relative ordering for ND tensors, so we need to save and add
+  // the non-updated start as an offset i.e. the second row of a 3x3 tensors
+  // starts at element 3 but sorter's second row only contains 0, 1, or 2
+  const int64_t orig_start = start;
+  while (start < end) {
+    const int64_t mid = start + ((end - start) >> 1);
+    const input_t mid_val = sort ? bd[sort[mid] + orig_start] : bd[mid];
+    if (!(mid_val >= val)) {
+      start = mid + 1;
+    } else {
+      end = mid;
+    }
+  }
+  return start;
+}
+
+// customized upper_bound func to ensure we can properly handle a sorter
+// argument std::upper_bound can not be used here since its customized
+// comparator requires both arguments to have the same type, which wouldn't
+// happen when comparing val of input_t to an indexer value from sorter of int64
+template <typename input_t>
+int64_t cus_upper_bound(
+    int64_t start,
+    int64_t end,
+    const input_t val,
+    const input_t* bd,
+    const int64_t* sort) {
+  // sorter gives relative ordering for ND tensors, so we need to save and add
+  // the non-updated start as an offset i.e. the second row of a 3x3 tensors
+  // starts at element 3 but sorter's second row only contains 0, 1, or 2
+  const int64_t orig_start = start;
+  while (start < end) {
+    const int64_t mid = start + ((end - start) >> 1);
+    const input_t mid_val = sort ? bd[sort[mid] + orig_start] : bd[mid];
+    if (!(mid_val > val)) {
+      start = mid + 1;
+    } else {
+      end = mid;
+    }
+  }
+  return start;
+}
+
+template <typename input_t, typename output_t>
+struct SearchsortedFunctor {
```

Comment:
If the structure implicates an operation in some kernel frame, we name it XXXFunctor. Like `AddFunctor` in `Loops` kernel frame. If the structure implicates a kernel frame, we name it XXXKernelFunctor. So here it should be the latter. We need name it `SearchsortedKernelFunctor`.